File: Programmers/lv.2/17680/solution.py
# https://programmers.co.kr/learn/courses/30/lessons/17680

def solution(cacheSize, cities):
    hit = 1; miss = 5
    if cacheSize == 0:
        return len(cities) * miss
    
    cities = [city.upper() for city in cities]
    cache = []
    time = 0
    for city in cities:
        if city in cache:
            time += hit
            cache.remove(city)
            cache.append(city)
        else:
            if len(cache) >= cacheSize:
                cache.pop(0)
            cache.append(city)
            time += miss
            
    return time

File: Programmers/lv.2/17680/Solution.java
import java.util.*;

class Solution {
    public int solution(int cacheSize, String[] cities) {
        if(cacheSize == 0) { return cities.length * 5; }

        for(int i = 0; i < cities.length; ++i) {
            cities[i] = cities[i].toLowerCase();
        }

        int answer = 0;
        List<String> cache = new LinkedList<>();
        for(int i = 0; i < cities.length; ++i) {
            String city = cities[i];
            
            if(cache.contains(city)) {
                cache.remove(city);
                cache.add(city);
                answer += 1;
            } else {
                cache.add(city);
                if(cache.size() > cacheSize) {
                    cache.remove(0);
                }

                answer += 5;
              }
        }

        return answer;
    }
}
